architecture/2024-11-ucp-component-threat-model.md

```diff
@@ -0,0 +1,259 @@
+
+**Status:** These mitigations are partial and require configuration by the operator. We will revisit and improve this area in the future.
+
+#### Threat: A Malicious Actor Could Intercept Unauthenticated Communication to Manipulate Data
+
+**Description:** Communication between the UCP and resource providers is currently unauthenticated. This means that there are no mechanisms in place to verify the authenticity of the communication between these components.
+
+**Impact:** Unauthenticated communication increases the risk of unauthorized access and actions. Attackers could potentially intercept or manipulate the communication between the UCP and resource providers, leading to data breaches or unauthorized operations.
+
+**Mitigations:**
+
+1. **Secure Communication Between UCP and Resource Providers:** Implement a form of authentication (e.g., mTLS) to verify the authenticity of the communication between the UCP and resource providers.
+2. **Network Policies and Firewall Rules:** Apply Kubernetes Network Policies to control traffic flow to and from the UCP. Configure firewalls to allow only necessary traffic between trusted components.
+
+**Status:** None of the mitigations is currently active and we created action items to work on them.
+
+#### Threat: A Malicious Actor Could Exploit SHA-1 Weaknesses to Generate Hash Collisions
+
+**Description:** A malicious actor could exploit the known vulnerabilities of the SHA-1 hashing algorithm to generate hash collisions. The UCP currently uses SHA-1 for hashing resource IDs and generating ETags. Although SHA-1 is not used in security-sensitive contexts, its vulnerabilities could still be exploited to create hash collisions, potentially leading to data integrity issues.
+
+**Impact:** Even though SHA-1 is used for non-security purposes, its vulnerabilities could be exploited to create hash collisions. This could result in data integrity issues, such as incorrect resource identification (can happen in ETag generation) or versioning (when we are hashing resource IDs to do some comparison). Using a stronger cryptographic hashing algorithm is essential to ensure the integrity and reliability of the system.
+
+**Mitigations:**
+
+1. **Replace SHA-1 with a Stronger Algorithm:**
+   1. **Action:** Identify all instances where SHA-1 is used in the codebase.
+   2. **Implementation:** Replace SHA-1 with SHA-256 or another secure hashing algorithm for hashing resource IDs and generating ETags. Ensure that the new algorithm is consistently used across all components. Test to verify that the change does not impact system functionality.
+
+**Status:** The mitigation is not active as of now. An action item has been created to update the cryptographic algorithm used in hashing resource IDs and generating ETags, as well as in other components of Radius.
+
+## Open Questions
+
+## Action Items
+
+1. **Use a Stronger Hashing Algorithm:**
```

Comment:
I don't understand this as an action item based on the document. The document says both that SHA-1 is used for non-cryptographic purposes, and also that it's a security threat. 

It feels like you're taking a compromise position that's going to result in a lot of work for nebulous benefit. 

Here's what I'd suggest:

- Let's review [this threat](https://github.com/radius-project/design-notes/pull/73/files#diff-de92b57845f019bf07433840060635c493191b42971c4601f782cd98757d8c03R227) for veracity.
- Let's get feedback from some security experts if we're unsure. 

We should really believe in the things we add as action items.

Reply:
Removed this one also since it is not really a security issue. Also removed the action item. We can revisit if we get another feedback.